Daniel datetimepicker

dkrasner opened the pull request:
This PR does two things:
* refactoring of playground using the `SplitView` cell/component for a cleaner layout
* adding a datetime (range) picker as requested in [this](https://github.com/APrioriInvestments/nativepython/issues/117)

## Approach
The playground refactoring is an exercise in using `SplitView` and somewhat cleaning up the layout code. 

Datetime picker cell/component follows our basic structure of server/cell client/component setup. It's a simple cell that allows the programmer to declare a start, or a start + end value in the case of a range picker. When a value is selected the js component sends over currently selected values over the WS in the following format:
```
{
event: 'change', 
target_id: ID,
start: VALUE,
end: VALUE
}
```

All datetime values going over the WS are unix time (seconds from epoch). 

## How Has This Been Tested?
This has been tested in playground interactively.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

braxtonmckee committed:
Basic infrastructure for a cells-test playground.

Adding uninitializableservice back in

basic datetimepicker

	modified:   object_database/frontends/object_database_webtest.py
	modified:   object_database/web/CellsTestService.py
	modified:   object_database/web/cells/__init__.py
	modified:   object_database/web/cells/cells.py
	new file:   object_database/web/cells_demo/datatimepicker.py
	modified:   object_database/web/content/ComponentRegistry.js
	new file:   object_database/web/content/components/DateTimePicker.js
	modified:   object_database/web/content/package.json

rebasing over dev

braxtonmckee commented:
I merged a bunch of your and eric's changes into dev. Can you rebase? Hard
to see what's going on because I squashed the commits...

On Fri, Jun 28, 2019 at 2:30 PM codecov[bot] <notifications@github.com>
wrote:

> Codecov
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137?src=pr&el=h1>
> Report
>
> ❗️ No coverage uploaded for pull request base (dev@cb2fe9d). Click here
> to learn what that means
> <https://docs.codecov.io/docs/error-reference#section-missing-base-commit>
> .
> The diff coverage is 40%.
>
> [image: Impacted file tree graph]
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137?src=pr&el=tree>
>
> @@          Coverage Diff           @@
> ##             dev     #137   +/-   ##
>
> ======================================
>
>   Coverage       ?   64.52%
>
> ======================================
>
>   Files          ?      103
>
>   Lines          ?    12089
>
>   Branches       ?        0
>
> ======================================
>
>   Hits           ?     7801
>
>   Misses         ?     4288
>
>   Partials       ?        0
>
> Impacted Files
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137?src=pr&el=tree> Coverage
> Δ
> object_database/web/cells/__init__.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxscy9fX2luaXRfXy5weQ==> 100%
> <100%> (ø)
> object_database/web/cells/cells.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxscy9jZWxscy5weQ==> 64.74%
> <18.75%> (ø)
> object_database/web/cells/views/split_view.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxscy92aWV3cy9zcGxpdF92aWV3LnB5> 25.8%
> <25.8%> (ø)
> ...ject_database/frontends/object_database_webtest.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL2Zyb250ZW5kcy9vYmplY3RfZGF0YWJhc2Vfd2VidGVzdC5weQ==> 30.35%
> <33.33%> (ø)
> object_database/web/CellsTestService.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9DZWxsc1Rlc3RTZXJ2aWNlLnB5> 33.8%
> <33.8%> (ø)
> object_database/web/cells_demo/datatimepicker.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxsc19kZW1vL2RhdGF0aW1lcGlja2VyLnB5> 40%
> <40%> (ø)
> object_database/web/cells_demo/split_view.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxsc19kZW1vL3NwbGl0X3ZpZXcucHk=> 64.7%
> <64.7%> (ø)
> object_database/web/cells_demo/card.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxsc19kZW1vL2NhcmQucHk=> 66.66%
> <66.66%> (ø)
> object_database/web/CellsTestPage.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9DZWxsc1Rlc3RQYWdlLnB5> 71.42%
> <71.42%> (ø)
> object_database/web/cells_demo/buttons.py
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137/diff?src=pr&el=tree#diff-b2JqZWN0X2RhdGFiYXNlL3dlYi9jZWxsc19kZW1vL2J1dHRvbnMucHk=> 71.42%
> <71.42%> (ø)
> ------------------------------
>
> Continue to review full report at Codecov
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137?src=pr&el=continue>
> .
>
> *Legend* - Click here to learn more
> <https://docs.codecov.io/docs/codecov-delta>
> Δ = absolute <relative> (impact), ø = not affected, ? = missing data
> Powered by Codecov
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137?src=pr&el=footer>.
> Last update cb2fe9d...3b6a471
> <https://codecov.io/gh/APrioriInvestments/nativepython/pull/137?src=pr&el=lastupdated>.
> Read the comment docs <https://docs.codecov.io/docs/pull-request-comments>
> .
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBC3XSFWKMC55DYQIU3P4ZKGFA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY2275Y#issuecomment-506834935>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBBHRUOH3NFPMRXADBDP4ZKGFANCNFSM4H4HSXXA>
> .
>

braxtonmckee commented:
OK, this looks cool. We need to be able to find out when the value changes. We should be able to pass it a `Slot` to hold the value, and if no slot is specified, it should allocate its own.  For instance, it would be nice to do something like

```
def cell(self):
   start = Slot(time.time())
   end = Slot(time.time())
   
   return cells.DateTimePicker(start=start, end=end) + cells.Subscribed(lambda: "Difference is %f" % (end.get() - start.get())")
```

I like the ability to use arrow keys to move around in the picker. It would be great if when I click and drag on 'month', every few pixels of vertical mouse movement moves the month up.

braxtonmckee commented:
Also, the playground reorg looks good. I think on the label, it might be better to make some other kind of labeled container cell we could use to do. It's pretty common to build little grids like this, so really we just need a convenience method. That might also make this code simpler because really I think the component we want is just the datetime picker - if the range picker is just two datetime pickers, we can get that by composing cells.

Thoughts?

dkrasner commented:
I'll get the rebase rolling

At the moment `onChange` the following json gets sent over the wire:
```
{
event: 'change', 
target_id: ID,
start: VALUE,
end: VALUE
}
```
can't the server listen and respond to that? or am i missing something

If I understand correct you would like a separate label component so for example you would defined a labelled date picker with something like `DatePicker(start=start, label_start=Label(text)` - correct? If so I could add this to the PR. 

Yes technically the the date range picker is 2 date pickers, but the convenience is two-fold: the api is simpler, i.e. the python programmer is just thinking of one cell/component and the data comes back in a single structure over the the wire. But I can easily revert back to a (single) data picker which is what i had first. Just let me know.

braxtonmckee commented:
On the messaging: there's no way for the user of the 'cells' API to
directly listen to messages.  The intended use is that a single Cell
completely encapsulates the relationship between the data layer
(object_database, Slots, and SessionState) and the UI, and that it
translates user actions on the UI into changes in the state model.  For
instance, when you create a Button, you pass in a lambda function that gets
called when the button gets clicked.  The lambda is free to write whatever
values it wants into the object database. In the case of the DatePicker,
the way this Cell should communicate to us is through a Slot - it should
display what's in the Slot and write changes back to the Slot. See
`SingleLineTextBox` for an analagous cell.

For the moment, I think all you need to do is to allow the user to pass in
a Slot object that contains the current value, and override the cell's
onMessage to write the newly selected value into the slot. If they pass a
float, you can make your own Slot and initialize it with the float.

I think the single datepicker is fine for the moment since we can build
whatever we need out of it on the python side, and if we need more complex
logic (e.g. first date has to be before second date) that can be enforced
serverside through the data binding.

On the label thing, yeah, I guess I'm saying that the DatePicker doesn't
need a label at all.  The act of labeling a DatePicker or a Button or any
other small piece of UI should be done through composing the DatePicker
with other cells.

On Tue, Jul 2, 2019 at 10:57 AM dkrasner <notifications@github.com> wrote:

> I'll get the rebase rolling
>
> At the moment onChange the following json gets sent over the wire:
>
> {
> event: 'change',
> target_id: ID,
> start: VALUE,
> end: VALUE
> }
>
> can't the server listen and respond to that? or am i missing something
>
> If I understand correct you would like a separate label component so for
> example you would defined a labelled date picker with something like DatePicker(start=start,
> label_start=Label(text) - correct? If so I could add this to the PR.
>
> Yes technically the the date range picker is 2 date pickers, but the
> convenience is two-fold: the api is simpler, i.e. the python programmer is
> just thinking of one cell/component and the data comes back in a single
> structure over the the wire. But I can easily revert back to a (single)
> data picker which is what i had first. Just let me know.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBG43UDG6CU4AQKGPZLP5NUEBA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBRX2A#issuecomment-507714536>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBBROMZD6IFIFSTASSLP5NUEBANCNFSM4H4HSXXA>
> .
>

On Tue, Jul 2, 2019 at 10:57 AM dkrasner <notifications@github.com> wrote:

> I'll get the rebase rolling
>
> At the moment onChange the following json gets sent over the wire:
>
> {
> event: 'change',
> target_id: ID,
> start: VALUE,
> end: VALUE
> }
>
> can't the server listen and respond to that? or am i missing something
>
> If I understand correct you would like a separate label component so for
> example you would defined a labelled date picker with something like DatePicker(start=start,
> label_start=Label(text) - correct? If so I could add this to the PR.
>
> Yes technically the the date range picker is 2 date pickers, but the
> convenience is two-fold: the api is simpler, i.e. the python programmer is
> just thinking of one cell/component and the data comes back in a single
> structure over the the wire. But I can easily revert back to a (single)
> data picker which is what i had first. Just let me know.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBG43UDG6CU4AQKGPZLP5NUEBA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZBRX2A#issuecomment-507714536>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBBROMZD6IFIFSTASSLP5NUEBANCNFSM4H4HSXXA>
> .
>

dkrasner committed:
coming back to a single, no label,  datetime picker

dkrasner committed:
adding Slot to DatePicker

updating playground Datetimepicker example

fixing typo in datetimepicker callback

minor updates to datetimepicker cell

Fixup for the DateTimePicker

dkrasner commented:
this all makes sense and I've taken care of most of it in the last few commits (will test more thoroughly when I have a moment in the coming days) 

the things about labels: visually it's just an html element with some text in and you could just have whatever `span` `div` etc and place it where. The advantage of declaring a label as part of an active element such as a `button`, `input` etc is that then these can be grouped together and explicitly associated using the `for` attribute. This allows for user event based effects on both the active element and the label, such as highlight on hover or some such.  See more [here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#Attributes). Since ids are generated automatically server side you would not be able to make sure a label is associated to the proper element (from what I can tell). 

just something to think about. Just let me know and I'll go from there.

braxtonmckee commented:
interesting on the label 'for' point.  i'll have to think more about that.

On Wed, Jul 3, 2019 at 10:46 AM dkrasner <notifications@github.com> wrote:

> this all makes sense and I've taken care of most of it in the last few
> commits (will test more thoroughly when I have a moment in the coming days)
>
> the things about labels: visually it's just an html element with some text
> in and you could just have whatever span div etc and place it where. The
> advantage of declaring a label as part of an active element such as a
> button, input etc is that then these can be grouped together and
> explicitly associated using the for attribute. This allows for user event
> based effects on both the active element and the label, such as highlight
> on hover or some such. See more here
> <https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#Attributes>.
> Since ids are generated automatically server side you would not be able to
> make sure a label is associated to the proper element (from what I can
> tell).
>
> just something to think about. Just let me know and I'll go from there.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBHGBRVGB4Q7BYPK5Z3P5S3VRA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZEV7WQ#issuecomment-508125146>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBBSYQR7HEQ7JDGGDNDP5S3VRANCNFSM4H4HSXXA>
> .
>

dkrasner commented:
@braxtonmckee you want me to break this up into two separate PR's, one dealing with refactoring of the playground and the other with the Datetimepicker (while you think about the label stuff)

dkrasner committed:
fixing missing export for Timestamp

fixing datetime input value format bug and removing all divStyle for DateTimePicker

fixing lint error

dkrasner commented:
@braxtonmckee i fixed the minor bug that broke `Timestamp`


the UX for Datetimepicker (or any input element) is pretty erratic due to the continual client event -> server slot -> "new" data to component.  I am going to have to think about how to make this work now with the current setup. But ideally we have have a situation where data goes over the wire when something changes and something on the other side of the wire needs to know about it, and only then.

dkrasner commented:
@braxtonmckee I fixed some of the input format issues, but the remaining focus input actions (for example, keeping focus on minutes and updating with arrow keys) is not going to work in the current setup and is best handled by a framework outside of the components themselves that deals with creation, update, and destruction of cells/components/nodes. 

This is something @darth-cheney  is working on right now and it should be out as a PR in the near future. 

So if this is all working for you, aside form the above, I say we merge it. Then merge Eric's work, and see if this all smoothes out.

braxtonmckee commented:
I just merged eric's PR. Let's rebase this on top of that and see if you can get it to update more smoothly now that we don't have to rebuild the cell from scratch every time.

dkrasner committed:
adding missing import un object_database_webtest.py

dkrasner committed:
fixing props look for timestamps

dkrasner committed:
fixing lifecycle related bugs

dkrasner committed:
fixing data flow and event handling issues for DatetimePicker

dkrasner commented:
@braxtonmckee  i rebased and tracked down various other bugs, and also sorted out how to make this all work right with the lifecycle and Slots frameworks

there is still some replacement bug that is floating around that I can't quite understand

on a bigger issue. The data flow (with slots) breaks html5 focusing (as I mentioned above) and that makes the UX of this input element pretty crappy. I am going to try to figure out how to capture the focus on a state and have it be consistent over the life of Component (this is in theory now possible with Eric's last PR)

dkrasner commented:
I don't see a way to control the focusing (element by element) in the html5 datetime-local type input element. 

Basically if we want to have html 5 UX we need a unidirectional data flow, i.e. we can't be resetting the values from the server all the time, which means no Slots or at least no Slots in their current implementation. 

There are two options: either live with the current, somewhat awkward in my view, datepicker. Or make our own, which will take a bit of time to get right. 

of course the data flow framework will be something we have to work around moving forward in other implementations

@braxtonmckee let me know what you think here

braxtonmckee commented:
What happens if, when you get an update from the server, you check if the
value matches what's in the element and just drop the message and don't
rebuild the component. I can certainly modify the slots mechanism to not
force a cell rebuild when it's a message from the cell itself that changes
the slot. Or is the problem that you're getting messages with lag and so
it's hard to tell whether to roll back?

On Wed, Jul 17, 2019 at 6:14 AM dkrasner <notifications@github.com> wrote:

> I don't see a way to control the focusing (element by element) in the
> html5 datetime-local type input element.
>
> Basically if we want to have html 5 UX we need a unidirectional data flow,
> i.e. we can't be resetting the values from the server all the time, which
> means no Slots or at least no Slots in their current implementation.
>
> There are two options: either live with the current, somewhat awkward in
> my view, datepicker. Or make our own, which will take a bit of time to get
> right.
>
> of course the data flow framework will be something we have to work around
> moving forward in other implementations
>
> @braxtonmckee <https://github.com/braxtonmckee> let me know what you
> think here
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBEI23FDXBLIJWYAKVTP73WJ5A5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2DXH4I#issuecomment-512193521>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBB3SOKO7QUNBTJLGPLP73WJ5ANCNFSM4H4HSXXA>
> .
>

dkrasner committed:
removing divStyle from Card and CardTitle

darth-cheney commented:
@braxtonmckee 

> I can certainly modify the slots mechanism to not force a cell rebuild when it's a message from the cell itself that changes the slot.
  
This might be a good interim solution, but it would then introduce a new (and perhaps unexpected, elsewhere) behavior to the Cell/Slot relationship
  
> you check if the value matches what's in the element and just drop the message and don't rebuild the component.
  
It's entirely possible to have a component conditionally render on an update given some criteria, though we might want to add that as some pre-render lifecycle on components for all update messages (something like `componentShouldRender()` which returns a boolean). The problem with this approach is that now we don't know across the board what the dataflow direction is between the server and the hierarchy of components. It will be an even more important issue when we leave the "child first" rendering strategy and actually render trees of components and their children (a re-render of a parent should call render on all its child components, and so on).
  
The problem we are experiencing is the confluence of several issues. One is that we are using the "input" component as a "display" component for data that gets shared across different Cells -- we want to be able to have some action elsewhere "change" the datetime in the picker that is not an action the user in question has taken on the picker itself. Should this not be the case, then the problem is easier to solve because the dataflow is as expected.
  
For now it might be best to do the following:
* Create a custom datetime picker from scratch, so we have complete control
* Ensure that input events on the picker do not fire any time a value changes, but rather wait
either for some time delay or for something like a `mouseleave` event -- this prevents a jerky experience when sending data back
* Indicate when a datetimepicker has been "synced" with the server, ie a change has been sent and received by the corresponding python Cell
* This way we won't lose focus and the specific component can deal with how it renders when new values come in
  
I think @dkrasner can get a from-scratch component like this up and running in a day or two. Let us know what you think

braxtonmckee commented:
OK, I think I have a protocol that make it easier for us to figure out
what's going on.

The idea is that both Slots and object_database have an internal clock that
always moves forward called a transaction_id.  Because of internal
implementation details, it'll probably be a pair of unsigned integers, but
it's guaranteed to increase each time something changes.  Each time I send
the client an update message, I can include that transaction_id. When you
have a current state, you'll know the transaction_id of the state that it's
showing. Whenever you send a message to the server, the server will respond
with the transaction_id associated with that change, and that message will
come across the wire _Before_ any cell updates. That way, after you trigger
a change, you can know exactly what transaction_id your change triggered.
Then, any updates that come back to you that are associated with cells
recomputing as of that transaction_id will be distinguishable from updates
that come from other parts of the system. I think that should be enough for
you to tell what's going on with the state management.

For instance, imagine you have two datetime pickers that share a slot (kind
of a stupid example, but it ought to work). Each one will have a
transaction_id, say 't0' associated with the single value they're showing.
When the user changes one, the client will send the server a message to
write into the slot. The server will roll the transaction_id forward to
't1' and respond to that update message with the new tid. The datepicker
that produced the update will be the only one that gets the message, and
it'll update its notion of the current state to 't1'. Then the server will
send out to _both_ cells the recalculated value, along with the
transaction_id of 't0'. The cell that triggered the change will recognize
its own update and know that it doesn't need to update itself, and the cell
that didn't will update, and update it's internal transaction_it to t1.  If
we fire off many transactions in quick succession, we're guaranteed that
the 'update transaction' message arrives before the state-change message,
so we shouldn't have this problem where several quick updates create
confusion.

This pattern should also alleviate the issues I've seen in the text editor.

Separately, on doing a from-scratch one, I think that's probably a good
idea, but lets make sure we fix all the recent regressions from the code
churn before we do that since I now have some broken deployments. But then,
yes, that seems like a good idea because there were some features we wanted
and it sounded like doing it in the existing component was going to be
infeasible.

@kdrasner what do you think?

On Wed, Jul 17, 2019 at 10:27 AM Eric Gade <notifications@github.com> wrote:

> @braxtonmckee <https://github.com/braxtonmckee>
>
> I can certainly modify the slots mechanism to not force a cell rebuild
> when it's a message from the cell itself that changes the slot.
>
> This might be a good interim solution, but it would then introduce a new
> (and perhaps unexpected, elsewhere) behavior to the Cell/Slot relationship
>
> you check if the value matches what's in the element and just drop the
> message and don't rebuild the component.
>
> It's entirely possible to have a component conditionally render on an
> update given some criteria, though we might want to add that as some
> pre-render lifecycle on components for all update messages (something like
> componentShouldRender() which returns a boolean). The problem with this
> approach is that now we don't know across the board what the dataflow
> direction is between the server and the hierarchy of components. It will be
> an even more important issue when we leave the "child first" rendering
> strategy and actually render trees of components and their children (a
> re-render of a parent should call render on all its child components, and
> so on).
>
> The problem we are experiencing is the confluence of several issues. One
> is that we are using the "input" component as a "display" component for
> data that gets shared across different Cells -- we want to be able to have
> some action elsewhere "change" the datetime in the picker that is not an
> action the user in question has taken on the picker itself. Should this not
> be the case, then the problem is easier to solve because the dataflow is as
> expected.
>
> For now it might be best to do the following:
>
>    - Create a custom datetime picker from scratch, so we have complete
>    control
>    - Ensure that input events on the picker do not fire any time a value
>    changes, but rather wait
>    either for some time delay or for something like a mouseleave event --
>    this prevents a jerky experience when sending data back
>    - Indicate when a datetimepicker has been "synced" with the server, ie
>    a change has been sent and received by the corresponding python Cell
>    - This way we won't lose focus and the specific component can deal
>    with how it renders when new values come in
>
> I think @dkrasner <https://github.com/dkrasner> can get a from-scratch
> component like this up and running in a day or two. Let us know what you
> think
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBGA6II57W5K6MNWGLTP74T6PA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2EMHZQ#issuecomment-512279526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBGWLHWAFK37OVNNTALP74T6PANCNFSM4H4HSXXA>
> .
>

dkrasner committed:
removing color method from Octicon and related cell/component changes

dkrasner committed:
adding Text cell to playground and removing divStyle from it

dkrasner committed:
refactoring style of Sequence

updaing CollapsiblePanel styling

minor code editor styling update

darth-cheney committed:
Adding PageView Cell/Component, with styling, and making it the primary view in ActiveWebService

darth-cheney committed:
Fixing unused import
Fixing linting issues

dkrasner commented:
Are you proposing to add another type of message that informs the client/component the corresponding transaction id that some event generated?

for example:
suppose you have a Datepicker and it has a `this.last_transaction_i d = null`
a datetime is selected, triggering a call back to the server
this triggers a Slot/ObDB update which then
sends a message to the client with the `transaction_id` and `Datepicker.last_transaction_id = transaction_id`

something like that?

braxtonmckee commented:
I think I would send the current transaction_id with all of the messages,
since it's always known.

On Fri, Jul 19, 2019 at 10:35 AM dkrasner <notifications@github.com> wrote:

> Are you proposing to add another type of message that informs the
> client/component the corresponding transaction id that some event generated?
>
> for example:
> support you have a Datepicker and it has a this.last_transaction_i d =
> null
> a datetime is selected, triggering a call back to the server
> this triggers a Slot/ObDB update which then
> sends a message to the client with the transcation id
>
> something like that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBC7MHK47IWWQKVA7JLQAHGJNA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2LZ7GI#issuecomment-513253273>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBB66VDSU4ADEK56XX3QAHGJNANCNFSM4H4HSXXA>
> .
>

dkrasner commented:
ok maybe I am missing something but how does a component know whether a transaction id correspond to a client event callback or some server side state change. 

with the same example:
suppose you have a Datepicker and it has a this.last_transaction_i d = null
a datetime is selected, triggering a call back to the server
this triggers a Slot/ObDB update with some `transaction_id`
the next update message comes with this `transction_id` and now the client/component needs to decide: is this the transaction id corresponding to its last event or is it a transaction id corresponding to some other server side event, or a client side event but for a diff session

braxtonmckee commented:
I'd have to make some changes to get this working, but the proposed flow is:

The datepicker has transaction_id = whatever it was when it got created.

> a datetime is selected, triggering a call back to the server
>
Imagine you select the datepicker and change it. You'll fire a message to
the server.  You include a guid to identify the state change.  The intent
of this message is to change some state on the server.  Server opens a
transaction, and notes the transaction_id which is unique and monotonically
increasing. It responds to your state-change message telling you the
transaction_id and the guid you originally sent.

> this triggers a Slot/ObDB update with some transaction_id
>
Yes, it delivers your message to the python version of the cell, who then
makes state changes either in object db or slots or both. We then
recalculate the state of the system allowing _only_ the changes due to your
update. This may involve propagating state changes in the slots and odb
infrastructure and recalculating new cells.  Any messages the python
version of the cell sends to the client will have this unique
transaction_id. Similarly, any new cells will be tagged with this
transaction_id.

>
> the next update message comes with this transction_id and now the
> client/component needs to decide: is this the transaction id corresponding
> to its last event or is it a transaction id corresponding to some other
> server side event, or a client side event but for a diff session
>
The _only_ messages with the transaction_id = the one the server responded
to your message with will be due to the state changes associated with
processing your message. I'll have to do some work to ensure that we don't
have race conditions when lots of things are happening at once. but in
principle you should be able to distinguish the state changes triggered by
your message uniquely and discard them.

Does that sound like it will work?


> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/APrioriInvestments/nativepython/pull/137?email_source=notifications&email_token=AB6OHBFOVQD7YFFHLXQCKI3QAHNXFA5CNFSM4H4HSXXKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2L7S5Y#issuecomment-513276279>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB6OHBAEI33GRTXHSHNZRZDQAHNXFANCNFSM4H4HSXXA>
> .
>

dkrasner committed:
removing div style from Subscribed and Subscribed sequence

dkrasner committed:
removing divStyle from Expands, Plot, Sheet, and CodeEditor

dkrasner committed:
removing divStyle from columns

dkrasner committed:
adding bundle

dkrasner committed:
removing remaining divStyles in components

dkrasner commented:
perhaps, but i am not 100% sure the added complexity is warranted, since at the least we can work with the current bi-directional data flow (even if it is not "standard" or advisable from the standpoint of many _reactive_ systems) by simply making our own components in these few special cases.

dkrasner committed:
adding latest js bundle

dkrasner committed:
basic custom datetime picker, fully functional

dkrasner committed:
checking out fixed version of Timestamp.js and updated cells_demo

dkrasner committed:
working version of the slider component for datetime picker

dkrasner committed:
linking slider events and datetimepicker input update and callbacks

dkrasner committed:
cleanup of datetimepicker and adding inpine documentation

dkrasner committed:
adding proper input date time type handling for the slider ranges

dkrasner commented:
@braxtonmckee check it out. You can input values manually (as in any datetime input), but if you doubleclick inside any input element a slider will appear. It sets appropriate values based on what you click on (year, month, etc)

(i don't account for different month lengths and leap years at the moment, but otherwise it's pretty close)

dkrasner committed:
adding a number of days in month checker for slider values, and minor cleanup of DatetimePicker

dkrasner committed:
adding datetimepicker-component-tests

dkrasner committed:
adding slider value setting to match the current datetime value; adding related datetimepicker tests

dkrasner commented:
@braxtonmckee as long as you like the UX i think this is good to go. It now has a proper days of month handling, nicer slider value setting to match the actual date and tests for the component.

braxtonmckee commented:
I just checked this out. The interaction with the slots framework looks correct, so that's good. On my browser, it lays out like this:

![image](https://user-images.githubusercontent.com/8184708/62218752-0eabcc00-b37b-11e9-9dfd-161b26b3d512.png)

which is not very compact. Can we get it all on one line? Also, it would be easier to edit the text if it were just a single text box, so that arrow keys move around. If the date is no longer properly formatted, OK to roll it back.  To retain the behavior of the slider, perhaps you can tell which part of the string they click and use that to figure out whether the slider should move the year, hour, etc.

Also, on the slider, i had been thinking that instead of creating an actual slider you can see, you could double click on a part of the datetime (say the month), and then just drag up and down, with no additional ux element displayed. In that mode, if you kept dragging up, so that the month goes past 12, then it should roll back to 1 and increment the year. Does that sound doable?

dkrasner commented:
the above screenshot looks weird and I am not seeing it at all on any of the screens here, or by playing around with the browser size

this is what I see:

![Screen Shot 2019-07-31 at 4 21 42 PM](https://user-images.githubusercontent.com/1154326/62219870-5d735900-b3af-11e9-8a22-9691fa25f3e1.png)



maybe you didn't rebuild the js bundle, or the browser is not hard-refreshed?


Making this into a single `input` element brings us back to version 0 above. We will lose focus, with the data flow, and won't have individual datetime subpart event control. Unless I am missing something in order to do this we will have to make a new html5 input element type which is certainly doable but I would need to figure out what that actually takes (it will take more than what we have done so far that I am pretty sure of).  You can use the `tab` key to move right. 

The "no-slider" slide/drag might be possible, but I would have to think about it. You might be able to do something like, make these (individual) input elements html5 `draggable`, then register the drag coordinates but instead of actually dragging the element on the screen change the input value depending on which direction the "drag" is occurring in. I could think about this but it would take a bit of experimenting? (using a single input as suggested about this would not be possible from what i can tell). If you issue is the slider taking up space that can be an overlay vertical/horizontal, etc

what do you think?

dkrasner commented:
perhaps with the chrome upgrade some of these issues will be solved. 

@braxtonmckee what do you think about the next step here, given my last comment

braxtonmckee commented:
This is part of APrioriInvestments/object_database
